mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java

```diff
@@ -688,10 +697,23 @@ public MessageResultIterator getMessages(MessageRange set, FetchGroup fetchGroup
 
     }
 
-    protected Map<MessageUid, MessageMetaData> deleteMarkedInMailbox(final MessageRange range, MailboxSession session) throws MailboxException {
-
+    protected Map<MessageUid, MessageMetaData> deleteWithPreHooks(final MessageRange range, MailboxSession session) throws MailboxException {
         final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        List<MailboxMessage> messagesWillBeDeleted = ImmutableList.copyOf(
+            messageMapper.findInMailbox(mailbox, range, FetchType.Metadata, -1));
+
+        DeleteOperation deleteOperation = DeleteOperation.from(
+            messagesWillBeDeleted.stream()
+                .map(mailboxMessage -> MetadataWithMailboxId.from(mailboxMessage.metaData(), mailboxMessage.getMailboxId()))
+                .collect(Guavate.toImmutableList()));
+
+        return Flux.fromIterable(preDeletionHooks)
+            .flatMap(hook -> hook.notifyDelete(deleteOperation))
+            .then(Mono.fromSupplier(Throwing.supplier(() -> deleteMarkedInMailbox(range, messageMapper))))
+            .block();
+    }
 
+    private Map<MessageUid, MessageMetaData> deleteMarkedInMailbox(MessageRange range, MessageMapper messageMapper) throws MailboxException {
```

Comment:
`deleteMarkedInMailbox()` should take `messagesWillBeDeleted` parameter but it is a task from Refactoring expunge PR. Wait for the cherry-pick